win32ss/drivers/miniport/bochs/bochsmp.c

```diff
@@ -0,0 +1,752 @@
+            {
+                RequestPacket->StatusBlock->Status = ERROR_INSUFFICIENT_BUFFER;
+                return FALSE;
+            }
+            return BochsQueryCurrentMode(DeviceExtension,
+                                         (PVIDEO_MODE_INFORMATION)RequestPacket->OutputBuffer,
+                                         RequestPacket->StatusBlock);
+        }
+
+        case IOCTL_VIDEO_RESET_DEVICE:
+        {
+            VideoDebugPrint((Info, "BochsStartIO - Reset device\n"));
+            return BochsResetDevice(DeviceExtension,
+                                    RequestPacket->StatusBlock);
+        }
+
+        case IOCTL_VIDEO_GET_CHILD_STATE:
+        {
+            VideoDebugPrint((Info, "BochsStartIO - Get child state\n"));
+            if (RequestPacket->OutputBufferLength < sizeof(ULONG))
+            {
+                RequestPacket->StatusBlock->Status = ERROR_INSUFFICIENT_BUFFER;
+                return FALSE;
+            }
+            return BochsGetChildState(DeviceExtension,
+                                      (PULONG)RequestPacket->OutputBuffer,
+                                      RequestPacket->StatusBlock);
+        }
+
+        default:
+        {
+            VideoDebugPrint((Warn, "BochsStartIO - Unknown IOCTL - 0x%08x\n",
+                             RequestPacket->IoControlCode));
+            break;
+        }
+    }
+
+    return FALSE;
+}
+
+VP_STATUS NTAPI
+BochsSetPowerState(
+    _In_ PVOID HwDeviceExtension,
+    _In_ ULONG HwId,
+    _In_ PVIDEO_POWER_MANAGEMENT VideoPowerControl)
+{
+    return NO_ERROR;
+}
+
+VP_STATUS NTAPI
+BochsGetPowerState(
+    _In_ PVOID HwDeviceExtension,
+    _In_ ULONG HwId,
+    _Out_ PVIDEO_POWER_MANAGEMENT VideoPowerControl)
+{
+    return ERROR_DEVICE_REINITIALIZATION_NEEDED;
+}
+
+VP_STATUS NTAPI
+BochsGetVideoChildDescriptor(
+    _In_ PVOID HwDeviceExtension,
+    _In_ PVIDEO_CHILD_ENUM_INFO ChildEnumInfo,
+    _Out_ PVIDEO_CHILD_TYPE VideoChildType,
+    _Out_ PUCHAR pChildDescriptor,
+    _Out_ PULONG UId,
+    _Out_ PULONG pUnused)
+{
+    PBOCHS_DEVICE_EXTENSION DeviceExtension = HwDeviceExtension;
+
+    VideoDebugPrint((Info, "Bochs: BochsGetVideoChildDescriptor Entry\n"));
+
+    if (ChildEnumInfo->Size < sizeof(VIDEO_CHILD_TYPE))
+        return VIDEO_ENUM_NO_MORE_DEVICES;
+
+    if (ChildEnumInfo->ChildIndex == 0)
+    {
+        /* Ignore ACPI enumerations */
+        return VIDEO_ENUM_INVALID_DEVICE;
+    }
+
+    *pUnused = 0;
+    if (ChildEnumInfo->ChildIndex == DISPLAY_ADAPTER_HW_ID)
+    {
+        *VideoChildType = VideoChip;
+        return VIDEO_ENUM_MORE_DEVICES;
+    }
+
+    if (ChildEnumInfo->ChildIndex != 1)
+        return VIDEO_ENUM_NO_MORE_DEVICES;
+
+    *UId = 0;
+    *VideoChildType = Monitor;
+
+    if (pChildDescriptor &&
+        ChildEnumInfo->ChildDescriptorSize >= VBE_EDID_SIZE &&
+        !DeviceExtension->IoPorts.RangeInIoSpace)
+    {
+        memcpy(pChildDescriptor,
```

Comment:
RtlCopyMemory

Reply:
As RtlCopyMemory is a #define for memcpy, I'll keep memcpy to not include <ndk/rtlfuncs.h>